be/src/exec/workgroup/work_group.cpp

```diff
@@ -112,19 +120,75 @@ void WorkGroupManager::destroy() {
+    add_metrics(wg);
     if (_workgroup_versions.count(wg->id()) && _workgroup_versions[wg->id()] == wg->version()) {
         return _workgroups[unique_id];
     } else {
         return get_default_workgroup();
     }
 }
 
+void WorkGroupManager::add_metrics(const WorkGroupPtr& wg) {
+    std::call_once(_init_metrics, []() {
+        StarRocksMetrics::instance()->metrics()->register_hook("work_group_metrics_hook",
+                                                               [] { WorkGroupManager::instance()->update_metrics(); });
+    });
+
+    if (_wg_metrics.count(wg->name()) == 0) {
+        //cpu limit
+        std::unique_ptr<starrocks::DoubleGauge> resource_group_cpu_limit_ratio(new DoubleGauge(MetricUnit::PERCENT));
+        //cpu concurrent
+        std::unique_ptr<starrocks::DoubleGauge> resource_group_cpu_use_ratio(new DoubleGauge(MetricUnit::PERCENT));
+        //mem limit
+        std::unique_ptr<starrocks::IntGauge> resource_group_mem_limit_bytes(new IntGauge(MetricUnit::BYTES));
+        //mem concurrent
+        std::unique_ptr<starrocks::IntGauge> resource_group_mem_allocated_bytes(new IntGauge(MetricUnit::BYTES));
+
+        StarRocksMetrics::instance()->metrics()->register_metric("resource_group_cpu_limit_ratio",
+                                                                 MetricLabels().add("name", wg->name()),
+                                                                 resource_group_cpu_limit_ratio.get());
+        StarRocksMetrics::instance()->metrics()->register_metric("resource_group_cpu_use_ratio",
+                                                                 MetricLabels().add("name", wg->name()),
+                                                                 resource_group_cpu_use_ratio.get());
+        StarRocksMetrics::instance()->metrics()->register_metric("resource_group_mem_limit_bytes",
+                                                                 MetricLabels().add("name", wg->name()),
+                                                                 resource_group_mem_limit_bytes.get());
+        StarRocksMetrics::instance()->metrics()->register_metric("resource_group_mem_allocated_bytes",
+                                                                 MetricLabels().add("name", wg->name()),
+                                                                 resource_group_mem_allocated_bytes.get());
+
+        _wg_cpu_limit_metrics.emplace(wg->name(), std::move(resource_group_cpu_limit_ratio));
+        _wg_cpu_metrics.emplace(wg->name(), std::move(resource_group_cpu_use_ratio));
+        _wg_mem_limit_metrics.emplace(wg->name(), std::move(resource_group_mem_limit_bytes));
+        _wg_mem_metrics.emplace(wg->name(), std::move(resource_group_mem_allocated_bytes));
+    }
+    _wg_metrics.emplace(wg->name(), wg->unique_id());
+}
+
+void WorkGroupManager::update_metrics() {
+    for (auto& wg_metric : _wg_metrics) {
+        WorkGroupPtr wg = _workgroups[wg_metric.second];
```

Comment:
here _workgroups[wg_metrics.second] is a destructive operation if the corresponding workgroup are removed. so use _workgroups.find(wg_metrics.second] instead of operator[].

Reply:
I'm so sorry, it's my fault, this is my first time writing c++ code, many principles are not very clear, I will change it as you said